### ISSUE TITLE
Feature: Enable specifying scheduling priority of nix builds

### ISSUE THREAD

**Issue by chreekat** [NixOS/nix#3020]
In multi-user Nix, builds are performed by the nix-daemon, and aren't part of the process tree that initiated the build. This means that `nice nix build` has no meaningful effect.

Perhaps nix could read getpriority(2) and pass that info to the daemon?

**Comment by Mic92** [comment]
Stil an issue.

**Comment by mikepurvis** [comment]
I would love to see prioritization built into `nix build` as well. It exists in Hydra, but not for CLI builds, and that's a bummer.

**Comment by magneticflux-** [comment]
For (re)building large derivation trees, a simple way to increase system utilization might be to give the less-relied-upon derivations more priority and take priority from derivations that build in parallel. This would speed up "bottleneck" non-parallel builds that are preventing more builds from starting at the expense of "leaf node" parallel builds that can be postponed.

I've been experimenting with large builds on a highly parallel server and I've noticed parallel builds will alternate between:
- being CPU-saturated by short-lived parallel builds `x`, `y`, `z`, and ignoring `A` for 30 seconds
- being stalled by a crucial package `A` (not dependent on `x`, `y`, or `z`) building sequentially a few minutes

The overall build time could be reduced by prioritizing work on `A`, acting like a preemptive scheduler for a series-parallel graph.

Some more complex algorithms to find optimal schedules I glanced at:
- https://doi.org/10.1109/PDCAT.2008.54
- https://doi.org/10.1007/978-3-031-12597-3_7

**Comment by chreekat** [comment]
@magneticflux- thanks, that's interesting, but just to be clear it is unrelated to the subject of this ticket. I wasn't to be able to *throttle* Nix builds in some environments.

**Comment by mikepurvis** [comment]
Certainly one pathway to a solution here could be defining an API for Nix's build scheduler and then making it pluggable— you could have the default implementation just be a basic FIFO affair like we have now, or power users would have the option of specifying alternative ones that would try to prioritize derivations with lots of dependencies or allow preemption or defer/throttle ones that are heavily parallel, or whatever else.

Such a thing wouldn't even necessarily have to be a binary-linked plugin (though that would likely be easiest), it could even be a separate executable or server that is called with a JSON blob of the current queue, and emits a response saying what to build next.

Any design of this would also need to account for scheduling on remote builders (see #6696). Specifically, a difficulty arises with multiple clients all trying to use the same remote builder— the builder itself should have some say about how to order what has been put in its queue, but that info needs to propagate back to the client so it can decide to maybe schedule on a _different_ remote builder if the one it initially selected one is (or becomes) too busy.

**Comment by pwaller** [comment]
@mikepurvis (repeating the @chreekat's above comment) -- that's offtopic for this issue, though it sounds interesting, it belongs somewhere else.

@chreekat I had the same problem, I want to do builds but I don't want to disturb other users on the system.

There is a global  `nix.daemonCPUSchedPolicy` on a nixos machine, this has the effect of configuring the systemd unit under which nix runs. https://github.com/NixOS/nixpkgs/blob/b597bff1b64322690f553058b4ed4a3522bbbc4a/nixos/modules/services/system/nix-daemon.nix#L197. The idle priority looks suitable for my use case, but it does note that on a machine with no spare CPU, the daemon could get starved (and therefore not do an update for a vital configuration update). There is lots of discussion also on this pull request which removed the `nix.daemonNiceLevel` https://github.com/NixOS/nixpkgs/pull/138741 

Apparently, `nice` doesn't have the effect you would expect anyway! See for example this comment https://github.com/NixOS/nixpkgs/pull/138741#issuecomment-924955890 which references the manpage: ("Under group scheduling, a thread's nice value has an effect for scheduling decisions only relative to other threads in the same task group. ...)"

I agree though it would be nice (ehem) to set priority on a per-'nix build invocation' basis. I would like to be able to specify when a specific build should have an idle priority, and I would like to do it when I invoke nix for that build, as an ordinary user.

**Comment by chreekat** [comment]
Actually I think @mikepurvis is on-topic, he's just thinking about the problem from a more general perspective!
